### PR TITLE
Run ci with locked and newer version of servant

### PR DESCRIPTION
**Pull request by PhilTaken**
Problem: the ci should run with two version of servant but is currently limited to the locked version

Solution: add a new ci build to run all test with another hackage snapshot

Also rewrite ci to run in a nix flake to adhere to the newest standards

#### Related changes (conditional)

- [x] Tests
  - If I added new functionality, I added tests covering it.
  - If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- [x] Documentation
  - I checked whether I should update the docs and did so if necessary:
    - [README](/README.md);
    - Haddock;
    - [Examples](/examples/).

- [x] Public contracts
  - Any modifications of public contracts comply with the [Evolution
  of Public Contracts](https://www.notion.so/serokell/Evolution-of-Public-Contracts-2a3bf7971abe4806a24f63c84e7076c5) policy.
  - I added an entry to the [changelog](../tree/master/CHANGES.md) if my changes are visible to the users
        and
  - provided a migration guide for breaking changes if possible.

#### Stylistic guide (mandatory)

- [x] Style
  - My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
  - My code complies with the [style guide](../tree/master/docs/code-style.md).
  - Each commit of this pull request contains a description.
    - For new features, this description contains the use case for the new functionality.
    - For bug fixes, it describes both the attacked problem and a solution for it.
    - For dependency version changes description shortly enlists features and breaking changes of the new library version or contains a link to its changelog.

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        cp -r --no-preserve=mode ${projectSrc} $out
+        ${pkgs.lib.concatMapStrings ({ name, subdirectory }: ''
+          cp ${cabalFiles}/${name}.cabal $out/${subdirectory}
+        '') local-packages}
+      '';
+
+      src = (if resolver == null then cabalsrc else let
+        hlib = pkgs.haskell-nix;
+        allSnapshots = builtins.attrNames hlib.snapshots;
+        isLts = version: builtins.substring 0 3 version == "lts";
```
Comment (rvem) [inline flake.nix:55]: Is it used?

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+  outputs = { self, nixpkgs, flake-utils, haskell-nix, hackage, stackage, ... }@inputs:
+  flake-utils.lib.eachSystem ["x86_64-linux"] (system: let
+    pkgs = import inputs.nixpkgs {
+      inherit system;
+      overlays = [ inputs.haskell-nix.overlay ];
+    };
+
+    # run ci with resolver defined in stack.yaml and a newer resolver for servant 0.19
+    resolvers = [ null "lts-18.28" ];
```
Comment (rvem) [inline flake.nix:23]: AFAICS, `lts-18.28` has the same `servant-client-core` version as currently used `lts-18.21`, see https://www.stackage.org/lts-18.28

Comment (rvem) [reply flake.nix:23]: As an alternative, you can modify `stack.yaml` with `servant-client-core-0.19` extra dependency without modifying the resolver

Comment (PhilTaken) [reply flake.nix:23]: Would that play out well with all dependencies? I assumed servant-client-core would depend on newer versions of some packages that would not be available with the old resolver

Comment (PhilTaken) [reply flake.nix:23]: Adding both `servant-0.19` and `servant-client-core-0.19` to stack.yaml does seem to resolve correctly but leads to ci errors on the haskell side

> AFAICS, lts-18.28 has the same servant-client-core version as currently used lts-18.21, see https://www.stackage.org/lts-18.28

I had a typo in the flake, meant to put `19.28` :facepalm:  but that leads to another error saying that `text-format` cannot be resolved

File: flake.nix (hunk 1 above)
Comment (PhilTaken) [reply flake.nix]: Meant to put in a helpful assert but hackage.nix and stackage.nix have decent enough error messages, removed!

File: flake.nix (hunk 2 above)
Comment (PhilTaken) [reply flake.nix]: > As an alternative, you can modify stack.yaml with servant-client-core-0.19 extra dependency without modifying the resolver

I changed it over to that approach, seems more sensible. I have to admit I do not fully understand how dependencies are resolved outside the "main" resolver but it seems much more flexible than I had assumed

**Comment by rvem** [comment]
@Martoon-00, could you please take a look at the build failure?

**Comment by Martoon-00** [comment]
Oh, I guess the actual problem is that you picked a slice with `servant-client-core-0.19` & `servant-client-0.18`, and it does not work.

Probably it would be nice if you find an automatic way to include all the necessary `servant-*` packages.

**Comment by PhilTaken** [comment]
> Oh, I guess the actual problem is that you picked a slice with `servant-client-core-0.19` & `servant-client-0.18`, and it does not work.
> 
> Looks like my request in the YT ticket is not actually sensible in the provided form. Probably it would be nice if you find an automatic way to include all the necessary `servant-*` packages.

I tried a more recent snapshot that includes the newer version of servant but that would fail with being unable to resolve other packages, such as `text-format`.

Adding `servant-client-0.19` fixes it for now.

An automatic way should be possible :thinking:

**Comment by Martoon-00** [comment]
Hmm, really :thinking: 

Looks like we have `allow-newer: true` in the `stack.yaml`, just meh. I now removed it in `master`.

**Comment by PhilTaken** [comment]
> Hmm, really thinking
> 
> Looks like we have `allow-newer: true` in the `stack.yaml`, just meh. I now removed it in `master`.

well, that sounds like an interesting flag for sure :joy:

File: .buildkite/pipeline.yml [hunk 3]
```diff
@@ -1,14 +1,4 @@
...
-   - nix-build ci.nix -A packages.servant-util.tests.servant-util-test
-   - ./result/bin/servant-util-test
-
- - label: test servant-util-beam-pg
-   commands:
-   - nix-build ci.nix -A packages.servant-util-beam-pg.tests.servant-util-beam-pg-test
-   - ./result/bin/servant-util-beam-pg-test
+  - label: check
+    commands:
+      - nix flake check
```
Comment (rvem) [inline .buildkite/pipeline.yml:4]: A small nitpick, but let's use `-L` to always have complete build logs in CI

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      component-tests = pkgs.lib.fold pkgs.lib.mergeAttrs {} (builtins.catAttrs "tests" (builtins.attrValues comps));
+      testruns = pkgs.lib.mapAttrs' (name: drv: {
+        name = "${name}-${depname}";
+        value = pkgs.runCommandLocal "${name}-run" {} ''
+          ${drv}/bin/${name} && touch $out
+        '';
+      }) component-tests;
+    in builtins.listToAttrs (map (value: { name = "${value.name}-${depname}"; inherit value; }) all-components) // testruns;
+  in {
+    checks = pkgs.lib.fold pkgs.lib.mergeAttrs {} (pkgs.lib.mapAttrsToList make-tests extraDeps);
```
Comment (rvem) [inline flake.nix:103]: Previously tests were split on per-package basis, let's keep it this way, thus they'll be two jobs: to check `servant-util` and `servant-util-beam-ng`

**Review by rvem — APPROVED** [review]
Have a few nitpicks, but aside from that LGTM

File: flake.nix (hunk 4 above)
Comment (PhilTaken) [reply flake.nix]: alright, I thought building all derivations as tests would be a nice addition. looking at the test run, they seem to all be covered by just these 2 tests :+1:

File: .buildkite/pipeline.yml (hunk 3 above)
Comment (PhilTaken) [reply .buildkite/pipeline.yml]: of course, good catch!

File: flake.nix (hunk 4 above)
Comment (rvem) [reply flake.nix]: OTOH, these tests aren't that time-consuming, so apparently, the current approach is good as well. Tests are run sequentially and there is no mess in logs

Comment (rvem) [reply flake.nix]: So feel free to ignore my initial suggestion :see_no_evil:

Comment (PhilTaken) [reply flake.nix]: flake is cleaner like this :shrug:

**Review by rvem — APPROVED** [review]

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,101 @@
+{
+  description = "A very basic flake";
```
Comment (rvem) [inline flake.nix:2]: Oh wait :see_no_evil:

Comment (PhilTaken) [reply flake.nix]: :eyes:

**Pull request merged**